File: 0001-0500/0084. Largest Rectangle in Histogram.java
// Hard
// Stack
// Time:O(n), Space:O(n)
// https://leetcode.cn/problems/largest-rectangle-in-histogram/

import java.util.Stack;
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<Integer> stack = new Stack<>();
        int maxArea = 0, n = heights.length;
        // i <= n bcz we add a dummy end height '0' at the end
        for (int i = 0; i <= n; i++) {
            int h = (i == n ? 0 : heights[i]);
            while (!stack.isEmpty() && h < heights[stack.peek()]) {
                int height = heights[stack.pop()];
                int width = stack.isEmpty() ? i : i - stack.peek() - 1;
                maxArea = Math.max(maxArea, height * width);
            }
            stack.push(i);
        }
        return maxArea;
    }
}

File: 0001-0500/0084. Largest Rectangle in Histogram.py
# Time:O(n), Space:O(n)

from typing import List

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        max_area = 0

        for i in range(len(heights)):
            while stack and heights[i] < heights[stack[-1]]:
                h_idx = stack.pop()
                h = heights[h_idx]
                w = i if not stack else i - stack[-1] - 1
                max_area = max(max_area, h * w)

            stack.append(i)

        while stack:
            h_idx = stack.pop()
            h = heights[h_idx]
            w = len(heights) if not stack else len(heights) - stack[-1] - 1
            max_area = max(max_area, h * w)

        return max_area
    